staging/src/k8s.io/client-go/tools/portforward/tunneling_dialer.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package portforward
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/httpstream/spdy"
+	constants "k8s.io/apimachinery/pkg/util/portforward"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/transport/websocket"
+	"k8s.io/klog/v2"
+)
+
+const PingPeriod = 10 * time.Second
+
+// tunnelingDialer implements "httpstream.Dial" interface
+type tunnelingDialer struct {
+	url       *url.URL
+	transport http.RoundTripper
+	holder    websocket.ConnectionHolder
+}
+
+// NewTunnelingDialer creates and returns the tunnelingDialer structure which implemements the "httpstream.Dialer"
+// interface. The dialer can upgrade a websocket request, creating a websocket connection. This function
+// returns an error if one occurs.
+func NewSPDYOverWebsocketDialer(url *url.URL, config *restclient.Config) (httpstream.Dialer, error) {
+	transport, holder, err := websocket.RoundTripperFor(config)
+	if err != nil {
+		return nil, err
+	}
+	return &tunnelingDialer{
+		url:       url,
+		transport: transport,
+		holder:    holder,
+	}, nil
+}
+
+// Dial upgrades to a tunneling streaming connection, returning a SPDY connection
+// containing a WebSockets connection (which implements "net.Conn"). Also
+// returns the protocol negotiated (currently v2 portforward), or an error.
+// The passed "protocols" are ignored.
+func (d *tunnelingDialer) Dial(protocols ...string) (httpstream.Connection, string, error) {
```

Comment:
protocols passed in are currently `[portforward.k8s.io]`, hard-coded at https://github.com/kubernetes/kubernetes/blob/54bcbc3c750a4cf9f87d10b7a69b64e250d4495e/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L194

the protocol we return *has* to be one of the ones passed in

there's nothing specific about the portforward protocol in this tunneling connection, which is why I suggested passing all the protocols to the server with a SPDY/3.1+ prefix and letting the server decide what we should speak, then stripping that prefix off and returning the selected one to the caller

Reply:
We're using `v2.portforward.k8s.io` to select the `TunnelingHandler` instead of the legacy `UpgradeAwareHandler` 

https://github.com/kubernetes/kubernetes/blob/cb05a861a9a6c180f1fb8f04e853155ded3d5f29/staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go#L113

We'll need to take that into account.

Reply:
I updated the tunneling protocol to use the `SPDY/3.1+` prefix approach. Please let me know what you think.

Reply:
looks good, thanks